app_dart/lib/src/model/firestore/ci_staging.dart

```diff
@@ -108,49 +121,109 @@ class CiStaging extends Document {
-      }
-      newRemaining = remaining - 1;
-
-      // Choose what to update in the document; by default we want to record the new conclusion, but we only
-      // want to record `remaining - 1` if the original conclusion was `scheduled`.
-      if (kDefaultTaskStatus == fields[checkRun]!.stringValue) {
-        log.info('$logCrumb: setting remaining to $newRemaining and changing ${fields[checkRun]!.stringValue}');
-        fields[checkRun] = Value(stringValue: conclusion);
-        fields[kRemainingField] = Value(integerValue: '$newRemaining');
+      // GitHub sends us 3 "action" messages for check_runs: created, completed, or rerequested.
+      //   - We are responsible for the "created" messages.
+      //   - The user is responsible for "rerequested"
+      //   - LUCI is responsible for the completed.
+      // Completed messages are either success / failure.
+      // "remaining" should only go down if the previous state was scheduled - this is the first state
+      // that is written by the scheduler.
+      // "failed_count" can go up or down depending on:
+      //   previous == failure && current == success: +1
+      //   previous != failure && current == failure: -1
+      // So if the test existed and either remaining or failed_count is changed; the response is valid.
+      if (recordedConclusion == kScheduledValue && conclusion != kScheduledValue) {
+        // Guard against going negative and log enough info so we can debug.
+        if (remaining == 0) {
+          throw '$logCrumb: field "$kRemainingField" is already zero for $transaction / ${doc.fields}';
+        }
+        remaining = remaining - 1;
         valid = true;
+      }
+
+      if (recordedConclusion == kFailureValue) {
+        // Only rollback the "failed" counter if this is a successful test run,
+        // i.e. the test failed, the user requested a rerun, and now it passes.
+        if (conclusion == kSuccessValue) {
+          log.info('$logCrumb: conclusion flipped to positive - assuming test was re-run');
+          if (failed == 0) {
+            throw '$logCrumb: field "$kFailedField" is already zero for $transaction / ${doc.fields}';
+          }
+          valid = true;
+          failed = failed - 1;
+        }
       } else {
-        log.warning(
-          "$logCrumb: '$conclusion' already recorded for ${fields[checkRun]!.stringValue}, not updated $remaining",
-        );
-        newRemaining = remaining;
-        fields[checkRun] = Value(stringValue: conclusion);
-        valid = false;
+        // previous "conclusion" is either freshly scheduled or a success.
+        if (recordedConclusion == kSuccessValue) {
+          // The test succeeded, the user requested a rerun (why?), and now it fails.
+          log.info('$logCrumb: conclusion flipped to negative - assuming test was re-run');
+        } else {
+          // The test is a new failure
+          log.info('$logCrumb: test failed');
```

Comment:
`recordedConclusion` can be "scheduled", "success", or "failure", correct? Looking at the `if` branches that lead to this like, it seems `recordedConclusion` can only possibly be in "scheduled" state ("success" and "failure" lead to other branches). If so, how do we conclude that the test is a new failure at this point?

Reply:
It is only "scheduled" when we create the document, that is not a status message that is sent to use from Github. We only get "failed" and "success" from github because those are the bits used by LUCI -> Cocoon -> GitHub.

As above; I've removed these logs.